clang/lib/Driver/ToolChains/SYCL.cpp

```diff
@@ -324,32 +329,75 @@ static bool selectBfloatLibs(const llvm::Triple &Triple, const Compilation &C,
       }
     }
 
-    UseNative = false;
-
-    // Check for intel_gpu_pvc as the target
-    if (Arg *SYCLTarget = Args.getLastArg(options::OPT_fsycl_targets_EQ)) {
-      if (SYCLTarget->getValues().size() == 1) {
-        StringRef SYCLTargetStr = SYCLTarget->getValue();
-        if (SYCLTargetStr == "intel_gpu_pvc")
-          UseNative = true;
-      }
-    }
-
-    auto checkBF = [](StringRef Device) {
-      return Device.starts_with("pvc") || Device.starts_with("ats");
-    };
+    // We need to select fallback/native bfloat16 devicelib in AOT compilation
+    // targetting for Intel GPU devices. Users have 2 ways to apply AOT,
+    // 1). clang++ -fsycl -fsycl-targets=spir64_gen -Xs "-device pvc,...,"
+    // 2). clang++ -fsycl -fsycl-targets=intel_gpu_pvc,...
+    // 3). clang++ -fsycl -fsycl-targets=spir64_gen,intel_gpu_pvc,...
+    // -Xsycl-target-backend=spir64_gen "-device dg2"
 
     std::string Params;
     for (const auto &Arg : TargArgs) {
       Params += " ";
       Params += Arg;
     }
+
+    auto checkBF = [](StringRef Device) {
+      return Device.starts_with("pvc") || Device.starts_with("ats") ||
+             Device.starts_with("dg2") || Device.starts_with("bmg");
+    };
+
+    auto checkSpirvJIT = [](StringRef Target) {
+      return Target.starts_with("spir64-") || Target.starts_with("spirv64-") ||
+             (Target == "spir64") || (Target == "spirv64");
+    };
+
     size_t DevicesPos = Params.find("-device ");
-    if (!UseNative && DevicesPos != std::string::npos) {
+    // "-device xxx" is used to specify AOT target device, so user must apply
+    // -Xs "-device xxx" or -Xsycl-target-backend=spir64_gen "-device xxx"
+    if (DevicesPos != std::string::npos) {
       UseNative = true;
       std::istringstream Devices(Params.substr(DevicesPos + 8));
       for (std::string S; std::getline(Devices, S, ',');)
         UseNative &= checkBF(S);
+
+      // When "-device XXX" is applied to speicify GPU type, user can still
+      // add -fsycl-targets=intel_gpu_pvc..., native bfloat16 devicelib can
+      // only be linked when all GPU types specified support.
+      // We need to filter CPU and FPGA target here and only focus on GPU
+      // device.
+      if (Arg *SYCLTarget = Args.getLastArg(options::OPT_fsycl_targets_EQ)) {
+        for (auto TargetsV : SYCLTarget->getValues()) {
+          if (!checkSpirvJIT(StringRef(TargetsV)) &&
+              !StringRef(TargetsV).starts_with("spir64_gen") &&
```

Comment:
Comment states 'filter out CPU and FPGA' targets, but GPU is checked here as well, which would seamingly negate the `-device` check above.

Reply:
Hi, @mdtoguchi 
If we use following command:
clang++ -fsycl -fsycl-targets=spir64,spir64_fpga,spir64_gen,spir64_x86_64 -Xsycl-target-backend=spir64_gen "-device pvc" assert-debug.cpp
Although only GPU toolchain is checked here, the sycl target value got from "-fsycl-targets=" is a list including "spir64", "spir64_fpga", "spir64_gen", "spir64_x86_64" and we only need to care about "spir64_gen" and corresponding "-device XXX" argument, so I filtered "spir64", "spir64_x86_64", "spir64_fpga".
Thanks very much.